modules/contracts/src.sol/transferDefinitions/ParameterizedPayments.sol

```diff
@@ -0,0 +1,139 @@
+  /// @param encodedState encoded transfer parameters
+  /// @dev encoded parameter data contains the receiver address, the expiration, the UUID, and the rate
+  function create(
+    bytes calldata encodedBalance,
+    bytes calldata encodedState
+  ) external override view returns (bool) {
+    // Unpack balance & state
+    Balance memory balance = abi.decode(encodedBalance, (Balance));
+    TransferState memory state = abi.decode(encodedState, (TransferState));
+    Rate memory rate = state.rate;
+
+    // Receiver should not have an initial balance within the transfer
+    require(balance.amount[1] == 0, "Cannot create parameterized payment with nonzero recipient init balance");
+
+    // Parameterized payment checks
+    require(rate.deltaAmount >= 1, "Per-unit amount must be at least 1 wei");
+    require(rate.deltaTime >= 15, "Per-unit time must be at least 15 seconds");
+    require(state.expiration > block.timestamp + 3 days, "Expiration must be at least 3 days in the future.");
+
+    // Sanity checks
+    require(state.receiver != address(0x0), "Receiver address cannot be the zero address!");
+    require(state.UUID != bytes32(0), "UUID cannot be null.");
+
+    return true;
+  }
+
+  /// @notice Resolves a parameterized payment from the payer (creator) to the payee (resolver)
+  /// @param encodedBalance balance the creator/payer is putting into the transfer; same as in create
+  /// @param encodedState encoded transfer parameters
+  /// @param encodedResolver contains the resolution data (UUID, payment amount) and a signature of said data
+  function resolve(
+    bytes calldata encodedBalance,
+    bytes calldata encodedState,
+    bytes calldata encodedResolver
+  ) external override view returns (Balance memory) {
+    // Unpack balance, state, resolver, & resolverdata
+    Balance memory balance = abi.decode(encodedBalance, (Balance));
+    TransferState memory state = abi.decode(encodedState, (TransferState));
+    TransferResolver memory resolver = abi.decode(encodedResolver, (TransferResolver));
+    ResolverData memory data = resolver.data;
+
+    // Signature check
+    bytes32 hashedData = keccak256(abi.encode(data));
+    require(state.receiver == LibChannelCrypto.verifyChannelMessage(hashedData, resolver.payeeSignature));
+
+    // State & resolver UUID should match
+    require(state.UUID == data.UUID);
+
+    // Expiration check
+    require(block.timestamp < state.expiration);
+
+    // Rate should not be exceeded; multiply by large number to avoid precision errors
+    uint256 timeElapsed = block.timestamp.sub(state.start);
+    uint256 averageRate = data.paymentAmountTaken.div(timeElapsed.multiply(2**64));
+    uint256 allowedRate = state.rate.deltaAmount.div(state.rate.deltaTime.multiply(2**64));
+    require(averageRate <= allowedRate);
+
+    // Transfer the payment amount
+    balance.amount[0] = balance.amount[0].sub(data.paymentAmountTaken);
```

Comment:
should there be a require that the `paymentAmountTaken` is not larger than the original amount in the transfer?

i know technically safe math would save you here, but i do love my sanity checks

Reply:
@LayneHaber wouldn't that fail when the transfer tries to resolve with a higher balance than in the transfer itself?

Reply:
Confirmed with @ArjunBhuptani , but added in just to be safe!